Reviewed `a1345693` against `feature/registry-driven-rust-assembly`. Step 6 of #581.

## Reproduced

Workspace tests, strict Clippy, `cargo fmt --check`,
`RUSTDOCFLAGS='-D warnings' cargo doc --workspace --no-deps`. The committed
generated files are unchanged and the C smoke test is clean under
ASan/LSan/UBSan.

Spot-checking the umbrella accounting: the two `syn` walks left in
`prebindgen-registry` are both inside the
`#[cfg(any(test, feature = "testing"))]` helper, so "production code no longer
inspects generated Rust" holds. Dropping `FromIterator` is right — freezing now
reads the registry, and an assembly built without one would be a different
object wearing the same name.

## The capability seal is wider than its own doc claims

`RustWriter`'s documentation says it is

> minted at file assembly and at no other point, which is what keeps planning
> code from rendering

and its `compile_fail,E0624` doctest pins that an adapter cannot construct one.
That doctest still passes, because it fences *constructing* a writer. What
changed is *obtaining* one: `Assembly::writer` is `pub`, and JniGen keeps its
assembly on `JniGenerationPlan` for the generator's whole life after freezing —
which includes Kotlin emission.

So this compiles today, inside `kotlin_emit.rs`:

```rust
if let Some(g) = &self.generation {
    let w: &prebindgen_registry::RustWriter = g.assembly().writer();
    for h in crate::jni::collect_vec_build_helpers(self) {
        let _spelled = w.emit_source_type(&h.elem);
    }
}
```

Kotlin emission spelling Rust source types is exactly what the fences around
`RustWriter` exist to prevent, and nothing reports it. Before this PR the
capability existed only inside `write_rust`, handed to each `render`.

The fix is small, and it also removes the redundancy below. Making
`Assembly::writer` `pub(crate)` and letting
`assert_edges_cover_rendered_calls` take the writer from the assembly it was
already given:

```rust
pub fn assert_edges_cover_rendered_calls<A: RustArtifact>(
    assembly: &Assembly<A>,
    namespace: &str,
) {
    let emit = assembly.writer();
```

builds the workspace, passes every test, and leaves the generated files
byte-identical. With it, the snippet above fails at the same error code the
existing doctest pins:

```
error[E0624]: method `writer` is private
    --> prebindgen-jni/src/jni/kotlin_emit.rs:1991:69
```

Both adapters already pass `assembly().writer()` as that argument, so neither
loses anything — and the JNI call site stops naming the same expression twice.

## The evidence helper's writer can now disagree with the assembly's

Independently of the seal: `assert_edges_cover_rendered_calls` takes an `emit`
that the assembly now also owns, and the two are not required to match. The
registry's own tests pass `RustWriter::for_test()` against an assembly holding
`RustWriter::new(&empty_registry(), None)`. Those two are equal — an empty
registry yields the same empty module map and the same `crate` default — so
nothing is wrong today. But the names the check computes are supposed to be the
names emission will emit, and a mismatched writer makes it check different ones
while still passing.

## A doc comment moved to the wrong function, again

`Assembly::reaches` lost its doc block to `Assembly::writer`:

```rust
/// Whether an artifact answering for this identity reaches the file.
///
/// Reachability is read here rather than when the assembly was frozen: an
/// adapter may reach an artifact after adding it, and the file is what
/// settles the answer.
/// The rendering capability frozen with this assembly.
pub fn writer(&self) -> &crate::RustWriter {
```

`reaches` is left with none. This is the same defect as #591's
`plan_handle_destructors`, and the same cause: a new function inserted directly
above an existing one takes its documentation, with no blank line to mark the
seam. Twice in three PRs, and `cargo doc -D warnings` is blind to both, since
every intra-doc link still resolves.

The paragraph is worth keeping where it belongs — that reachability is settled
by the file rather than by freezing is the property
`an_artifact_reached_after_freezing_is_emitted` pins.

Nothing here blocks the PR. The seal is the one worth fixing before it lands,
since a doc comment claiming a fence that does not hold is what a later reader
will trust.

— Claude Code (Opus 5)

--- a/prebindgen-c/src/lib.rs
+++ b/prebindgen-c/src/lib.rs
@@ -351,11 +351,7 @@ impl Cbindgen {
         // dependency edges name every call its artifacts render — the
         // completeness the emission-time check reasons from.
         #[cfg(test)]
-        prebindgen_registry::write::assert_edges_cover_rendered_calls(
-            self.gen.assembly(),
-            self.gen.assembly().writer(),
-            "c",
-        );
+        prebindgen_registry::write::assert_edges_cover_rendered_calls(self.gen.assembly(), "c");
         Ok(prebindgen_registry::write::write_rust(
             self.gen.assembly(),
             out_path,

--- a/prebindgen-jni/src/jni/mod.rs
+++ b/prebindgen-jni/src/jni/mod.rs
@@ -907,12 +907,6 @@ impl JniGen {
                 .as_ref()
                 .expect("resolved JniGen has no frozen generation plan")
                 .assembly(),
-            self.decls
-                .generation
-                .as_ref()
-                .expect("resolved JniGen has no frozen generation plan")
-                .assembly()
-                .writer(),
             "jni",
         );
         Ok(prebindgen_registry::write::write_rust(

--- a/prebindgen-registry/src/write.rs
+++ b/prebindgen-registry/src/write.rs
@@ -160,13 +160,13 @@ impl<A: RustArtifact> Assembly<A> {
         self.artifacts.iter()
     }
 
-    /// Whether an artifact answering for this identity reaches the file.
-    ///
-    /// Reachability is read here rather than when the assembly was frozen: an
-    /// adapter may reach an artifact after adding it, and the file is what
-    /// settles the answer.
     /// The rendering capability frozen with this assembly.
-    pub fn writer(&self) -> &crate::RustWriter {
+    ///
+    /// Crate-private on purpose. `RustWriter` is minted at file assembly and
+    /// handed to each artifact as it renders; handing it out here would let
+    /// any holder of an assembly spell Rust source types outside that
+    /// boundary — Kotlin emission holds one for the generator's whole life.
+    pub(crate) fn writer(&self) -> &crate::RustWriter {
         &self.emit
     }
 
@@ -175,6 +175,11 @@ impl<A: RustArtifact> Assembly<A> {
         self.guards.iter()
     }
 
+    /// Whether an artifact answering for this identity reaches the file.
+    ///
+    /// Reachability is read here rather than when the assembly was frozen: an
+    /// adapter may reach an artifact after adding it, and the file is what
+    /// settles the answer.
     pub fn reaches(&self, key: &ArtifactKey) -> bool {
         self.provided
             .get(key)
@@ -271,18 +276,20 @@ impl<A: RustArtifact> AssemblyBuilder<A> {
 /// edges, and this checks the edges against what rendering actually emits.
 ///
 /// `namespace` is the adapter's operation namespace, the one it passes to
-/// [`RustWriter::operation_ident`](crate::RustWriter::operation_ident).
+/// [`RustWriter::operation_ident`](crate::RustWriter::operation_ident). The
+/// writer itself comes from the assembly, so the names this computes are the
+/// ones its emission will emit.
 ///
 /// # Panics
 ///
 /// Naming the caller, the callee and the undeclared call, if a rendered body
 /// calls a converter its artifact did not declare.
 #[cfg(any(test, feature = "testing"))]
-pub fn assert_edges_cover_rendered_calls<A: RustArtifact>(
-    assembly: &Assembly<A>,
-    emit: &crate::RustWriter,
-    namespace: &str,
-) {
+pub fn assert_edges_cover_rendered_calls<A: RustArtifact>(assembly: &Assembly<A>, namespace: &str) {
+    // The assembly's own writer, so the names checked here are the names its
+    // emission will emit. A separately supplied one could differ and the check
+    // would still pass, having examined different names.
+    let emit = assembly.writer();
     // What every artifact actually renders, which is the ground truth both
     // halves are checked against: a call resolves to an artifact by the name
     // that artifact defines, whether the identity is an operation or an

--- a/prebindgen-registry/src/write/tests.rs
+++ b/prebindgen-registry/src/write/tests.rs
@@ -297,7 +297,7 @@ fn honest_edges_pass_the_evidence_check() {
         }),
     ]);
 
-    assert_edges_cover_rendered_calls(&assembly, &crate::RustWriter::for_test(), "test");
+    assert_edges_cover_rendered_calls(&assembly, "test");
 }
 
 /// Claiming an identity the artifact does not render would satisfy
@@ -318,7 +318,7 @@ fn a_claimed_identity_must_be_rendered() {
         over_claims: true,
     })]);
 
-    assert_edges_cover_rendered_calls(&assembly, &crate::RustWriter::for_test(), "test");
+    assert_edges_cover_rendered_calls(&assembly, "test");
 }
 
 /// Two reachable artifacts under one identity are a planning error. A

